framework/src/wifi_manager/wifi_manager_dhcp_internal.c

```diff
@@ -0,0 +1,209 @@
+
+#ifndef CONFIG_WIFIMGR_DISABLE_DHCPS
+static dhcp_node_s g_dhcp_list = {WIFIMGR_IP4_ZERO, WIFIMGR_MAC_ZERO};
+#endif
+static uint8_t num_sta = 0;
+
+/**
+ * Internal DHCP server APIs
+ */
+void dhcps_inc_num(void)
+{
+	num_sta++;
+}
+
+void dhcps_dec_num(void)
+{
+	if (num_sta > 0) {
+		num_sta--;
+	}
+}
+
+void dhcps_reset_num(void)
+{
+	num_sta = 0;
+}
+
+uint8_t dhcps_get_num(void)
+{
+	return num_sta;
+}
+
+#ifndef CONFIG_WIFIMGR_DISABLE_DHCPS
+wifi_manager_result_e dhcps_start(dhcp_sta_joined cb)
+{
+	struct in_addr ip = {.s_addr = 0x012fa8c0 };
+	struct in_addr netmask = {.s_addr = 0x00ffffff};
+	struct in_addr gw = {.s_addr = 0x012fa8c0};
+
+	_dhcps_remove_list();
+	WIFIMGR_SET_IP4ADDR(WIFIMGR_SOFTAP_IFNAME, ip, netmask, gw);
+
+	if (dhcp_server_start(WIFIMGR_SOFTAP_IFNAME, cb) != OK) {
+		ndbg("[DHCP] DHCP Server - started fail\n");
+		return WIFI_MANAGER_FAIL;
+	}
+	nvdbg("[DHCP] DHCP Server - started successfully\n");
+
+	return WIFI_MANAGER_SUCCESS;
+}
+
+wifi_manager_result_e dhcps_stop(void)
+{
+	struct in_addr in = { .s_addr = INADDR_NONE };
+	WIFIMGR_SET_IP4ADDR(WIFIMGR_SOFTAP_IFNAME, in, in, in);
+
+	if (dhcp_server_stop(WIFIMGR_SOFTAP_IFNAME) != OK) {
+		ndbg("[DHCP] DHCP Server - stopped failed\n");
+	}
+	nvdbg("[DHCP] DHCP Server - stopped successfully\n");
+	return WIFI_MANAGER_SUCCESS;
+}
+
+dhcp_status_e dhcps_add_node(void *msg)
+{
+	dhcp_node_s *node = (dhcp_node_s *)msg;
+	if (node != NULL) {
+		nvdbg("[DHCP] IP: %d:%d:%d:%d\n",
+			  ((char *)&(node->ipaddr))[0],
+			  ((char *)&(node->ipaddr))[1],
+			  ((char *)&(node->ipaddr))[2],
+			  ((char *)&(node->ipaddr))[3]);
+		nvdbg("[DHCP] MAC: %02x:%02x:%02x:%02x:%02x:%02x\n",
+			  node->macaddr[0], node->macaddr[1],
+			  node->macaddr[2], node->macaddr[3],
+			  node->macaddr[4], node->macaddr[5]);
+	} else {
+		ndbg("[DHCP] No DHCP node exists\n");
+		return DHCP_ERR;
+	}
+
+	if (memcmp(g_dhcp_list.macaddr, node->macaddr, 6) == 0) {
+		g_dhcp_list.ipaddr = node->ipaddr;
+		return DHCP_EXIST;
+	}
+	g_dhcp_list.ipaddr = node->ipaddr;
+	memcpy(g_dhcp_list.macaddr, node->macaddr, 6);
+	return DHCP_OK;
+}
+
+void dhcps_del_node(void)
+{
+	//TODO: Currently only one client can be joined and left
+	g_dhcp_list.ipaddr = 0;
+	for (int i = 0; i < 6; i++) {
+		g_dhcp_list.macaddr[i] = 0;
+	}
+}
+
+static void _dhcps_remove_list(void)
+{
+	//TODO: Currently only one client can be joined and left
+	g_dhcp_list.ipaddr = 0;
+	for (int i = 0; i < 6; i++) {
+		g_dhcp_list.macaddr[i] = 0;
+	}
+}
+#endif //CONFIG_WIFIMGR_DISABLE_DHCPS
+
+
+/**
+ * Internal DHCP client APIs
+ */
+#ifndef CONFIG_WIFIMGR_DISABLE_DHCPC
+wifi_manager_result_e dhcpc_get_ipaddr(void)
+{
+	int ret;
+	struct in_addr ip;
+	wifi_manager_result_e wret = WIFI_MANAGER_FAIL;
+
+	ret = dhcp_client_start(WIFIMGR_STA_IFNAME);
+	if (ret != OK) {
+		WIFIADD_ERR_RECORD(ERR_WIFIMGR_CONNECT_DHCPC_FAIL);
+		ndbg("[DHCPC] get IP address fail\n");
+		return wret;
+	}
+
+	ret = netlib_get_ipv4addr(WIFIMGR_STA_IFNAME, &ip);
```

Comment:
I am not sure but... local struct in_addr may updated through this function(seems that bring address) but it didn't check any of validation, also it returns just result, not address.... is it enough?

Reply:
Here, the local 'struct in ip' is not used, but just fetching the ip address assigned by DHCP module. If we can get ip addr using netlib successfully, we might think the ip is successfully assigned. We cannot know the IP addr offered by the DHCP server in advance, actually there is nothing to do about it. Thanks!

Reply:
I am mentioning netlib_get_ipv4addr itself. you meaning null comes in addr through the ioctl in netlib_getipv4addr?

Reply:
In netlib_get_ipv4addr(),
if socket and/or ioctl fail, then it returns a ERROR or ioctl error return value with garbage ip address, and dhcpc_get_ipaddr() will return WIFI_MANAGER_FAIL in the end.
Therefore, checking the results of netlib_get_ipv4addr() can indirectly verify the ip is successfully obtained through DHCP or not.